Common Table Expression support added "out-of-the-box"

vlado opened the pull request:
### Summary

This PR adds `.with` query method which makes it super easy to build complex queries with Common Table Expressions.

It basically just "wraps" the passed arguments in the way `Arel::SelectManager.with` expects it. The biggest advantages this brings are: 

* You can just use standard ActiveRecord relations instead of manually building `Arel::Nodes::As` nodes
* It returns `ActiveRecord::Relation` so you don't loose any flexibility

See example below:

```ruby
Post.with(
  posts_with_comments: Post.where("comments_count > ?", 0),
  posts_with_tags: Post.where("tags_count > ?", 0)
)

# Replaces

posts_with_comments_table = Arel::Table.new(:posts_with_comments)
posts_with_comments_expression = Post.arel_table.where(posts_with_comments_table[:comments_count].gt(0))
posts_with_tags_table = Arel::Table.new(:posts_with_tags)
posts_with_tags_expression = Post.arel_table.where(posts_with_tags_table[:tags_count].gt(0))

Post.all.arel.with([
  Arel::Nodes::As.new(posts_with_comments_table, posts_with_comments_expression),
  Arel::Nodes::As.new(posts_with_tags_table, posts_with_tags_expression)
])
# WITH posts_with_comments AS (
#   SELECT * FROM posts WHERE (comments_count > 0)
# ), posts_with_tags AS (
#   SELECT * FROM posts WHERE (tags_count > 0)
# )
# SELECT * FROM posts
```

vlado committed:
`.with` query method added.

Construct common table expressions with ease and get `ActiveRecord::Relation` back.

kaspth commented:
@vlado @simi looks like you're doing the same work at the same time. Ref: https://github.com/rails/rails/pull/37973

Maybe you want to compare notes as well?

simi commented:
@kaspth not sure how could I miss that. I have pushed also fix for update_all/delete_all in my branch and it is missing recursive for now (to keep the PR simple).

@vlado would you mind if I port some of your changes into my PR, mark you as an author and continue in there? You have really nice documentation and also I can port some tests.

vlado commented:
Hi @simi and @kaspth thanks for the feedback. What a coincidence that we worked on same thing same time :)

I forgot to mention in PR description that I already added support for recursive expressions `Post.with(:recursive, cool_posts: "union to get cool posts...")`. It follows Arel API for recursive calls.  See more here: https://github.com/rails/rails/pull/37944/files#diff-25e9a621b650cd62a407c4ce7a117b20R77-R86.

I agree that we co-author this but I think it is easier to finish it here. From what I see only thing missing is `update_all / delete_all` support?

@kaspth What do you think?

simi commented:
@vlado the recursive api looks a little weird for me. I don't think there's any other AR relation method which uses "argument as a modifier of behaviour". Therefore I have decided in my PR to not introduce it for now to not block basic implementation and open another PR later once the first one get merged.

Comparing our code changes I have spotted those:

1. I have separated with logic to separate class (`ActiveRecord::With`). Would you mind to port them as well if you like that?

2. In `build_arel` method you have removed support for `Array` and for `Hash` your code is different. Can you explain on those changes? My idea was to not change logic/API to get 100% compatible with original `ctes_in_my_pg` gem to get easy transition.

Finally if you can prepare failing specs for `delete_all` and `update_all ` I can submit PR to your branch to support those relation calls as well.

vlado commented:
Hi @simi, thanks for the constructive feedback. Replies are below.

> @vlado the recursive api looks a little weird for me. I don't think there's any other AR relation method which uses "argument as a modifier of behaviour". Therefore I have decided in my PR to not introduce it for now to not block basic implementation and open another PR later once the first one get merged.

Hm, you could be right here. I also can not think of any AR relation method at this moment. I initially started with `with.recursive(...)` but when I checked how Arel works I saw that `.with(:recursive, ...)` is used and I liked it cause it is much simpler to implement. I was also thinking about `.with_recursive`.

> Comparing our code changes I have spotted those:
> 
> 1. I have separated with logic to separate class (`ActiveRecord::With`). Would you mind to port them as well if you like that?

Looks like a good idea 👍 

> 2. In `build_arel` method you have removed support for `Array` and for `Hash` your code is different. Can you explain on those changes? My idea was to not change logic/API to get 100% compatible with original `ctes_in_my_pg` gem to get easy transition.

Arrays are still supported but only if all items are `Arel::Nodes::As`. I didn't add documentation for this cause I didn't see options where you can pass Arel nodes documented anywhere (probably because Arel API is considered private).

Regarding Hash changes, I think there was an issue with `select = Arel::Nodes::SqlLiteral.new("(#{expression.to_sql})")` line in one of the tests in SQLite (I can check it later). Also, I tried to add meaningful error messages if unsupported arguments are passed.

I never used https://github.com/DavyJonesLocker/postgres_ext or https://github.com/kmurph73/ctes_in_my_pg in any of my apps (i was using my own patch) but I used them as inspiration when starting this PR.
That being said, I think I haven't break any compatibility with them in case of the upgrade, and even if we break some it should be super easy to upgrade. I think we should prioritise to make this best it can be.

> 
> Finally if you can prepare failing specs for `delete_all` and `update_all ` I can submit PR to your branch to support those relation calls as well.

Sure, I'll add failing tests.

simi commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -1195,6 +1251,37 @@ def build_select(arel)
+          when Hash then
+            with_value.map do |name, expression|
+              table = Arel::Table.new(name)
+              expression = expression.arel if expression.is_a?(ActiveRecord::Relation)
```
Any chance to keep logic in case/when here and support also `Arel::SelectManager`?

https://github.com/rails/rails/pull/37973/files#diff-bf6dd6226db3aab589916f09236881c7R1205-R1215

vlado replied:
`Arel::SelectManager` is supported. It is just passed as is. No need to wrap it with `Arel::Nodes::SqlLiteral`

simi replied:
Ahh, ok. What do you think to keep case/when logic instead of `if`s here?

vlado replied:
Ok, I'll check that when I'm back at my laptop later today.

My idea was that `expression` is either `Arel::SelectManager` or `Arel::Nodes:SqlLiteral` in the end. `ActiveRecord::Relation` is transformed to select manager with `arel` method and `String` into sql literal. I could revisit it to use case/when and even add unsupported argument error in else block then.

simi replied:
It just looks more readable IMHO. If unsupported expression is passed to `Arel::Nodes::As`, it will raise `TypeError` later during visiting that node. It will probably make sense to just allow supported in when/case and raise on unsupported to not let bubble up this low-level exception up to users.

vlado replied:
> It just looks more readable IMHO. 

I wanted to avoid nested case/when but I don't have strong opinion about this and agree with your proposal.

> If unsupported expression is passed to `Arel::Nodes::As`, it will raise `TypeError` later during visiting that node. It will probably make sense to just allow supported in when/case and raise on unsupported to not let bubble up this low-level exception up to users.

Yes, that was my idea also to show some meaningful error to the user instead of low-level one.

simi replied:
:handshake:

vlado replied:
I changed the code to use case/when and you were right, looks better now.
Only downside is that you need to know all Arel classes that can be passed (`Arel::SelectManager`, `Arel::Nodes::Union`, ...).

simi commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -1195,6 +1251,37 @@ def build_select(arel)
       end
 
+      def build_with(arel)
+        return if with_values.empty?
```
Is this needed? If `with_values` is empty, `with_values.map` will result into empty array. You just need to check in the end for `with_statements.any?` to apply `arel.with` if present.

vlado replied:
I think guard clause works great here.

It would also work without it but you need to add `any?` check anyways and all the calls before will be called also.

simi replied:
There is no guard clause used for another `build_*` methods, that is the idea to stick closer to current code.

simi commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -1195,6 +1251,37 @@ def build_select(arel)
+      def build_with(arel)
+        return if with_values.empty?
+
+        is_recursive = with_values.delete(:recursive)
```
I would prefer just `recursive` for this variable if you'll decide to keep recursive in this form of API.

vlado replied:
:+1: Changed to `recursive`

kaspth commented:
```ruby
Post.with(
  posts_with_comments: Post.where("comments_count > ?", 0),
  posts_with_tags: Post.where("tags_count > ?", 0)
)
```

I'm not too keen on this API. Seems verbose, yet still fairly low level. I haven't used CTEs before but just realized a case where I could, what's the use cases you've needed them for? And how do you act on the values? I'm guessing `posts_with_comments`/`posts_with_tags` are available to use in a custom SQL string?

What I'm saying is since Rails is an extracted, not built, framework, I'd prefer to see full production samples (with renames or some business logic stripped if needed). It seems like `with` is only half the equation, but there's also selecting on the resulting CTE. Plus what people can pass to `with` also includes a `UNION ALL`, I'd be curious to see those cases too. All in the name of figuring out how to cut and where to scope the API. Active Record isn't necessarily just Arel but nicer, it's free to be more geared. 😊

simi commented:
@kaspth CTEs are super useful since they work as a "temporary table". Our codebase is full of CTEs and this will make it much nicer, since we usually need to fallback to just executing SQL string written in heredoc. Also queries using CTE are often more readable. This will allow to compose those queries based on AR scopes instead of messing with SQL strings (we usually interpolate SQL string with relation.to_sql).

Yes, I agree all examples in this (and my PR) are really simple and not real reason to use CTE. I think there are articles around describing CTE and how/when to use them. On the other side Rails active record tests are full of examples not really making sense, but doing great job testing the feature.

I understand this not really well known feature of SQL (mainly for beginners). I have learned about that just few months ago. Introducing this feature with nice and friendly guide explaning purpose of this Active Record API will be great service to help people understanding CTEs.

> What I'm saying is since Rails is an extracted, not built, framework, I'd prefer to see full production samples (with renames or some business logic stripped if needed).

Just one quick example I do remember is we need to group and filter data and then join them with belongs_to relation. I'll try to prepare some examples with current code and how it will work with current pull request to show you some real examples extracted from real application.

ℹ️ This is also super useful in combination with recently introduced bulk upserts in combination with RETURNING (which did not make it yet to Rails codebase).

vlado commented:
Great questions @kaspth and you are guessing right :)

I'll try to explain with one example

## Example

At [BetterDoc](https://www.betterdoc.org) we have a 60+ lines SQL query that we use to dynamically calculate clinic rank based on some params (medical problem, geo location and radius, ...).
Keep in mind that this is a super simplified version that should be enough to explain the use case.

### Domain

We have 2 models in our domain. `Clinic` with `name` only, and `Case Number` which stores number of operations (`count`) clinic did for specific medical topic (`code`).

<img width="558" alt="Screenshot 2019-12-15 20 43 02" src="https://user-images.githubusercontent.com/5148/70867940-891c6780-1f7b-11ea-814b-f97a2ef618ef.png">

```ruby
class Clinic
  has_many :case_numbers
end

class CaseNumber
  belongs_to :clinic
end
```

### SQL Query

Here is a full SQL that is explain below.

```sql
WITH relevant_clinics AS (
    SELECT clinics.id, clinics.name, SUM(case_numbers.count) AS cases_count 
    FROM clinics INNER JOIN case_numbers ON case_numbers.clinic_id = clinics.id
    WHERE case_numbers.code ILIKE 'F4%'
    GROUP BY clinics.id 
),
stats AS (
    SELECT
        PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY cases_count DESC) AS quartile_2,
        PERCENTILE_CONT(0.50) WITHIN GROUP (ORDER BY cases_count DESC) AS quartile_1
    FROM relevant_clinics
)

SELECT name, cases_count,
    CASE 
        WHEN cases_count > (SELECT quartile_2 FROM stats) THEN 'High'
        WHEN cases_count > (SELECT quartile_1 FROM stats) THEN 'Average'
        ELSE 'Low' 
    END AS rank
FROM relevant_clinics
ORDER BY cases_count DESC
```

First step is to get clinics that match provided medical topic and store them as temporary result set  (`relevant_clinics`). Then we reference them to get some statistics from them and save those as another temporary results set (`stats`). We then fire main query which references both of them to get final results with rank and other stuff we need.

### How to organise this in Rails

Since ActiveRecord does not have support for CTE we started thinking how to organise this in our app.

We started by putting SQL into heredoc and executing it with `find_by_sql` and `count_by_sql`. This could work for some report or similar but it didn't work for us cause we needed `ActiveRecord::Relation` flexibility (chainability, lazy load, ...).

Next try was with `Arel.with` and it was a step forward but we were still missing `ActiveRecord::Relation` goodies. `Clinic.where(Clinic.arel_table[:id].in(arel_that_returns_clinic_ids))` "hack" was not an option cause we needed to keep order and other stuff from original query.

In the end I wrote a monkey patch that adds `.with` to `ActiveRecord::Relation` and we were able to write our query like this.

```ruby
class Clinic < ApplicationRecord
  has_many :case_numbers

  scope :matches_medical_topic, -> (q) { joins(:case_numbers).where("case_numbers.code ILIKE ? "#{q}%") }

  def self.search(q)
    relevant_clinics = matches_medical_topic(q)
      .select("cllinics.id, clinics.name, sum(case_numbers.count) AS cases_count
      .group("clinics.id")

    stats = select(
        "PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY cases_count DESC) AS quartile_2",
        "PERCENTILE_CONT(0.50) WITHIN GROUP (ORDER BY cases_count DESC) AS quartile_1"
      )
      .from("relevant_clinics")

    with(relevant_clinics: relevant_clinics, stats: stats)
      .from("relevant_clinics AS clinics")
      .order("cases_count DESC")
  end

  def self.with_rank
    select(:id, :name, :cases_count)
      .select("CASE WHEN cases_count > (SELECT quartile_2 FROM stats) THEN 'High' WHEN cases_count > (SELECT quartile_1 FROM stats) THEN 'Average' ELSE 'Low' END as rank")
  end
end
```

and use it like:

```ruby
Clinic.search("f4").with_rank # => ActiveRecord::Relation
Clinic.search("f4").count # => Integer
```

Since there are gems to do same thing I assumed that other people would like to see this in Rails so I started this PR.

## Conclusion

You can see from my example that I used `from("relevant_clinics AS clinics")` to be able to chain it with other relations. Alternative is to use join `joins("JOIN relevant_clinics ON relevant_clinics.id = clinics.id")`. 

I absolutely agree that this is a bit too verbose and low level but I haven't been able to figure out better API yet. I'm eager to hear your ideas and to make it more "geared" :)

Maybe the way to go could be to have something similar to associations and/or scopes that once defined will abstract everything for the user. Similar to the way `joins(:association_name)`works

```ruby
class Post
  has_many :case_numbers

  common_table :relevant_clinics, -> (q) {  join(:case_numbers).where("code = ?", q) }
  common_table :stats,-> { select(...).from(:relevant_clinics) }
end

Post.with(:relevant_clinics, :stats)
```

vlado commented:
I see now that I forgot about UNION example. You can find one in tests here https://github.com/rails/rails/pull/37944/files#diff-25e9a621b650cd62a407c4ce7a117b20R77-R86

Not a real life example but it is getting late here in Croatia :)

simi commented:
@vlado thanks for very detailed info. Our usecase and story is really similar, but in the end I have used https://github.com/kmurph73/ctes_in_my_pg as a `Relation#with` implementation.

@kaspth my idea was to introduce this `with` logic into AR step by step. Start with simple `with`, then continue with recursive if it will be needed. We can also later see how `with` is doing and if there are some new patterns like that "scope-like" `common_table` idea mentioned in @vlado 's comment.

I think demand for `WITH` support is not neglible. You can google for `cte rails active record site:stackoverflow.com` (sorry, I have no idea how to share search result link). There are also some blog post mentioning how to integrate CTEs into ActiveRecord.

Few examples including which strategy is used to integrate CTE to ActiveRecord (just selected from first search page):

* https://sonnym.github.io/2017/06/05/common-table-expressions-in-activerecord-a-case-study-of-quantiles/ (custom SQL in `select` and `from` + arel based solution)
* https://typefast.blog/common-table-expressions-within-rails/ (`find_by_sql` for ActiveAdmin)
* https://www.scimedsolutions.com/articles/75-arel-part-ii-common-table-expressions (arel based solution)
* https://hashrocket.com/blog/posts/recursive-sql-in-activerecord (mixing `to_sql` on scope and AR query methods, using recursive form of CTE)
* https://hashrocket.com/blog/posts/writable-common-table-expressions (using `connection.select_values`, explaining how to chain multiple insert/update/delete queries together, this PR is not really related to this use case; it will be really nice to support but it will need some time to bring some nice API for this; we use this pattern often)
* https://blog.castle.co/tech/10-06-2016/complex-sql-in-rails (using sql and `connection.select_all`, also using views to hide CTEs into DB - this approach is not usable if you need to build the WITH statement on the fly).

_PS: I was thinking to gemify this as well in some general way (not supporting only PostgreSQL), but monkeypatching of private methods from ActiveRecord and Arel would be needed. That was the main motivation to send my pull request to integrate this into core._

vlado commented:
> @vlado thanks for very detailed info. Our usecase and story is really similar, but in the end I have used https://github.com/kmurph73/ctes_in_my_pg as a `Relation#with` implementation.

I found out about `ctes_in_my_pg` late, postgres_ext was not working with Rails 6 so I did my own patch.

> I think demand for WITH support is not neglible.

Totally agree. We are using CTEs with Rails for more then 4 years now and it is a pain from the beginning. I was also giving talks about this topic (organising complex SQL queries in Rails) few times and from the feedback I got regarding CTE I was actually surprised that this was not implemented and decided to do it properly this time :)

kaspth commented:
> @kaspth my idea was to introduce this `with` logic into AR step by step. Start with simple `with`, then continue with recursive if it will be needed. We can also later see how `with` is doing and if there are some new patterns like that "scope-like" `common_table` idea mentioned in @vlado 's comment.

That's pretty much the opposite of what I'm curious to explore here or how I'd proceed. I'd rather we have as many real world cases to base this off of and extract the commonalities. Then if that looks like something that's worth it, we can consider adding that and some lower level constructs. As it is I don't think there's enough meat here to add `with` and I'm thinking of bowing out and passing to someone else on the team who feels more strongly about this.

For instance the `common_table` abstraction doesn't have much sway with me currently, from what I've gathered CTEs are for one off queries, so I'd personally rather put all those parts together.

> I think demand for `WITH` support is not neglible. You can google for `cte rails active record site:stackoverflow.com` (sorry, I have no idea how to share search result link). There are also some blog post mentioning how to integrate CTEs into ActiveRecord.

Sure, unfortunately that doesn't necessarily mean that we want to maintain something in Active Record directly. Then we'd be looking at APIs for gems instead, of which the current proposed `with` seems to fit with better.

@vlado I'm curious about your usage of `from` in your examples, can you expand on that? Particularly where you alias the `relevant_clinics` into `clinics`?

Another thing that's tough for me, although I like the `with_rank` split, is that `cases_count` isn't defined in there. So calling `with_rank` is dependent on the other method first.

simi commented:
@kaspth just try to take a look at this feature from different perspective.`WITH` is just another SQL clause, same as `WHERE`, `FROM`, `JOIN` or `ORDER` for example. `WITH` was introduced in SQL-99 standard and is widely supported by all "major" relation databases.

> Common table expressions are supported by Teradata, DB2, Firebird,[8] Microsoft SQL Server, Oracle (with recursion since 11g release 2), PostgreSQL (since 8.4), MariaDB (since 10.2), MySQL (since 8.0), SQLite (since 3.8.3), HyperSQL and H2 (experimental).[9] Oracle calls CTEs "subquery factoring".[10]

_https://en.wikipedia.org/wiki/Hierarchical_and_recursive_queries_in_SQL#Common_table_expression_

My motivation to support `WITH` in ActiveRecord is to make it possible to compose queries using `WITH` (for any reason) via nice ActiveRecord chainable API. The same you can do with custom joins, selects, wheres and orders today.

Sometimes you can avoid `SQL` oriented approach and just ask `ActiveRecord` for data like `Account.all.eager_load(:users)` and let the `ActiveRecord` do the best to build the query for you. Sometimes you need to use more `SQL` oriented approach and build chain of few custom selects, joins, wheres, ... and get the desired result. IMHO both approaches are totally valid. `ActiveRecord` excels in both of them already. Introducing `WITH` support will make the second approach even more flexible.

> Another thing that's tough for me, although I like the with_rank split, is that cases_count isn't defined in there. So calling with_rank is dependent on the other method first.

Yes, but I think this is something you need to handle carefully already if you use more `SQL` oriented approach. For example if you select or order by relation column, you need to ensure related table is properly joined.

> Sure, unfortunately that doesn't necessarily mean that we want to maintain something in Active Record directly. Then we'd be looking at APIs for gems instead, of which the current proposed with seems to fit with better.

Supporting this in gem is really hard since `ActiveRecord` internals are changing often and this needs to monkey patch various lower level APIs including AREL ones. There's solid support in AREL already for `WITH`. I'll extract my AREL change to make this work on `update_all`/`delete_all` as well to separated pull request (not introducing `Relation#with`, but just fixing this in AREL for `UpdateManager` and `DeleteManager`. Having this change merged will make it at least a little more easier to maintain in gemified form and AREL can benefit from that as well.

> For instance the common_table abstraction doesn't have much sway with me currently, from what I've gathered CTEs are for one off queries, so I'd personally rather put all those parts together.

> As it is I don't think there's enough meat here to add with and I'm thinking of bowing out and passing to someone else on the team who feels more strongly about this.

CTEs are not definitely useful just for one off queries. We use them for regular queries for various reasons already mentioned before. If you would like to not continue in this review I would like to thank you anyway for your time spent already on this. I can try to bring some attention here to get someone another to review if this feature is not actually finally rejected by that decision.

vlado commented:
> @vlado I'm curious about your usage of from in your examples, can you expand on that Particularly where you alias the relevant_clinics into clinics?

@kaspth I'm aliasing relevant_clinics as clinics here to be able to easily chain and call other query methods later. For example to do something like:

```ruby
Clinic.search("f4").with_rank.where(id: [1, 2, 3])
# WITH relevant_clinics AS (...) SELECT ... WHERE clinics.id IN (1,2,3) FROM relevant_clinics clinics  
```

As mentioned you could also use join `joins("JOIN relevant_clinics ON relevant_clinics.id = clinics.id")`

> Another thing that's tough for me, although I like the with_rank split, is that cases_count isn't defined in there. So calling with_rank is dependent on the other method first.

Agreed, I also did not liked it but it was good enough :) 

Really appreciate your participation on this PR and would not want you to quit. You asked great questions and I like your thinking.

I would say that biggest challenge (at least it was for us) is that you have expression(s) that builds your temporary table(s) and then you have multiple ways you want to use them (select without rank, select with rank, aggregate, call some additional where, ...) which makes it hard to put all this things together in one abstraction. That is way I proposed the `common_table` abstraction which is responsible just for the first part and then you are responsible to do something with it. Similar how joins works `CaseNumber.joins(:clinic).where("clinics.name ILIKE ?", "%query%")`. Not saying this is perfect solution but I'm more then happy to explore this further.

That being said I must also agree with @simi that just adding `with` method to `Relation` is already a huge improvement and it makes using CTE's with Rails 100 times better :)
Also, even when we find a proper abstraction and API  for common cases `with` should keep current functionality cause there will always be a case when you need to go low level to build your query.

PS. I'm sorry for a later reply. Was more focused on family stuff for holidays ☃️

vlado commented:
Hi @jeremy , just wanted to check if you are still interested to review and merge this? Or to take it further like @kaspth proposed? I'll be happy to discuss and do any additional work needed.

FYI I packaged this as a gem also in a meantime so we can use it in our projects https://github.com/vlado/activerecord-cte.

jeremy commented:
Let's wrap this up.

> I checked how Arel works I saw that .with(:recursive, ...) is used and I liked it cause it is much simpler to implement. I was also thinking about .with_recursive.

Using a positional symbol argument to change behavior does feel idiosyncratic. Recommend omitting support for the first PR, or introducing it as a distinct method as you suggest.

Regarding Array and Hash support that @simi mentioned: are you both happy with the API as currently exposed? Should we pull anything else over from #37973?

Regarding API and big-picture fitness for Active Record: I agree this is a "bottom up" feature, mapping a SQL capability to AR rather than the other way around. I see this as similar to `.joins(:accounts)` expecting other conditions on the joined table.

> Maybe the way to go could would be to have something similar to associations and/or scopes that once defined will abstract everything for the user. Similar to the way joins(:association_name)works. Option to go low level (current implementation) could also stay.

That's a great instinct and direction to pursue. I think it should be separate from this enabling work, however. We can make it possible and give a strong footing for other libraries (and PRs) to introduce such abstractions.

vlado commented:
Thanks @jeremy 

> > I checked how Arel works I saw that .with(:recursive, ...) is used and I liked it cause it is much simpler to implement. I was also thinking about .with_recursive.
> 
> Using a positional symbol argument to change behavior does feel idiosyncratic. Recommend omitting support for the first PR, or introducing it as a distinct method as you suggest.

Agreed, will switch to `.with_recursive`.
 
> Regarding Array and Hash support that @simi mentioned: are you both happy with the API as currently exposed?

I'm happy with the Array and Hash support and I do not expect any problems there (@simi concern for people that are upgrading from existing gems).

Should we pull anything else over from #37973?

I think we should pull `update_all` / `delete_all` functionality.
 
> Regarding API and big-picture fitness for Active Record: I agree this is a "bottom up" feature, mapping a SQL capability to AR rather than the other way around. I see this as similar to `.joins(:accounts)` expecting other conditions on the joined table.
> 
> > Maybe the way to go could would be to have something similar to associations and/or scopes that once defined will abstract everything for the user. Similar to the way joins(:association_name)works. Option to go low level (current implementation) could also stay.
> 
> That's a great instinct and direction to pursue. I think it should be separate from this enabling work, however. We can make it possible and give a strong footing for other libraries (and PRs) to introduce such abstractions.

Agreed 👍

vlado commented:
Hi @jeremy, please have a look on my changes when you gonna have time.

Few things to note:

#### String is not valid argument anymore

https://github.com/rails/rails/pull/39235 changed the way CTEs are built in Arel so it is not possible to pass string as a single argument anymore. It shouldn't be too hard to add support back but I do not think we need it.

```ruby
Post.with("posts_with_comments AS (SELECT * FROM posts WHERE comments_count > 0)")
# vs
Post.with(:posts_with_comments, "SELECT * FROM posts WHERE comments_count > 0")
```

#### `.with(:recursive, ...)` changed to `.with_recursive(...)`

Recursive is implemented as a distinct method now. It accepts 3 arguments (name, non recursive, recursive query) and builds union for you. I like this cause you can do everything with AR relations, no need to use Arel to build union or pass it as string.
Old way (passing hash/array argument) is still supported.

#### Didn't add `.update_all` and `.delete_all` support

I tried to port `.update_all` and `.delete_all` from https://github.com/rails/rails/pull/37973 but it will not be that simple. Implementation there doesn't cover all scenarios. For example `Post.with(:top_posts, ...).joins("JOIN top_posts ON top_posts.id = posts.id")` works and `Post.with(:top_posts, ...).from("top_posts AS posts")` does not work.
I just flagged `.with` and `.with_recursive` as invalid for `delete_all` and I suggest we continue work on this in other PR?

#### Security

If string is passed as common table name and/or expression it is automatically converted to sql literal. I do not see problem with this but wanted to check with you.

For name it is required or passing columns with name will not work.

```ruby
Comment.with_recursive("replies(id, parent_id, depth)", "SELECT ...", "SELECT ...")
# replies(id, parent_id, dept) will be quoted if not converted to SQL literal.
# WITH RECURSIVE "replies(id, parent_id, depth)" AS (SELECT ... UNION ALL SELECT ...)
```

dark-panda commented:
We're a big fan of CTEs where I'm at, and we currently use the ctes_in_my_pg gem to provide this functionality (https://github.com/kmurph73/ctes_in_my_pg). The interface here appears to be the same, with one difference that I recently upstreamed to ctes_in_my_pg, which is the addition of allowing the `MATERIALIZED` and `NOT MATERIALIZED` PostgreSQL extensions which were added in PostgreSQL 12 (https://www.postgresql.org/docs/12/sql-select.html).

ctes_in_my_pg is not compatible with 6.1.0.rc1 due to the recent changes in the `with` functionality in ActiveRecord, but if this feature is being added to Rails core then I'd prefer to direct my energy here to help with this effort. Is this functionality being considered for Rails 6.1.0? I realize that we're in the RC stage so perhaps it's too late for 6.1.0, but I can deal with a local monkey patch for this in the meantime if it's being considered for, say, 6.2.0. To that end, would the addition of the aforementioned Postgres-specific behaviour be useful, or would such an extension be out-of-scope for Rails?

vlado commented:
@dark-panda Since current implementation accepts `Arel::Nodes::As` it shouldn't be hard to make materialization work already.

Something like this should work:

```ruby
class MyModel < ApplicationRecord
  def self.with_not_materialized(name, relation)
    table = Arel::Table.new(name)
    expression = "NOT MATERIALIZED (#{relation.to_sql})"
    with(Arel::Nodes::As.new(table, expression))
  end
end
```

I've extracted initial work in this PR to the gem (https://github.com/vlado/activerecord-cte) that is compatible with Rails 6.

zzak commented:
@vlado How are things going with the gem? Would you be interested in picking this back up? :bow:

vlado commented:
> @vlado How are things going with the gem? Would you be interested in picking this back up? 🙇

Hi @zzak, gem is doing great, thanks for asking 😃

I'm happy to continues with this. Let me check what has changed in the meantime and see what needs to be changed to resolve conflicts and make CI green again.

JuceMan reviewed (approved):
>   #@``

vlado commented:
> @vlado How are things going with the gem? Would you be interested in picking this back up? 🙇

Hi @zzak, I've rebased and added few more tests. Everything looks good to me. Let me know what you think?

mildred commented:
What's missing to get this merged?

Having CTE available in rails would be a major improvement.

simi commented:
@vlado are you still interested in this PR? Would you mind rebasing?

simi commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -303,6 +303,151 @@ def _select!(*fields) # :nodoc:
+    #   # )
+    #   # SELECT comments.* FROM replies
+    #
+    # or pass them as a `Hash` but be aware that `ActiveRecord::Relation` does not have `UNION` support
```
I was thinking about introducing UNION support ActiveRecord, this could be great reasoning.

casperisfine commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -303,6 +303,151 @@ def _select!(*fields) # :nodoc:
+    #
+    #   Post
+    #     .with(:posts_with_comments, Post.where("comments_count > ?", 0))
+    #     .with(:posts_with_tags, Post.where("tags_count > ?", 0))
```
Same here. Having both `.with(:posts_with_tags, Post.where("tags_count > ?", 0))` and `Post.with(posts_with_comments: Post.where("comments_count > ?", 0))` doesn't seem very valuable but adds lots of complexity.

I think it would be better to focus on a single way to do CTE.

simi replied:
:+1: IMHO it will be nice to support hash-like interface and be able to specify multiple CTEs in one `with` call.

```ruby
Post.with(
  posts_with_comments: Post.where("comments_count > ?", 0),
  posts_without_comments: Post.where("comments_count = ?", 0)
)
```

vlado replied:
Yes, that makes most sense.

vlado replied:
We support only hash-like interface now.

casperisfine reviewed (commented):
I wonder if @kamipo has opinions about this feature.

kbrock commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -303,6 +303,151 @@ def _select!(*fields) # :nodoc:
+    #   # )
+    #   # SELECT * FROM posts_with_tags AS posts
+    #
+    #   Post.with(:posts_with_tags, Post.where("tags_count > ?", 0).joins("JOIN posts_with_tags ON posts_with_tags.id = posts.id"))
```
I think the parens on this are wrong. Maybe:

```suggestion
    #   Post.with(:posts_with_tags, Post.where("tags_count > ?", 0)).joins("JOIN posts_with_tags ON posts_with_tags.id = posts.id")
```

B. It is too bad we have to resort to this joins clause.
I say this because it feels like I will want to join the primary model to the CTE for most (all?) of my uses of CTE. Not sure if it is the same for you.

Is it possible to implement:

```
Post.with(:posts_with_tags, Post.where("tags_count > ?", 0)).joins(:posts_with_tags)
```

vlado replied:
I think it is possible and I proposed some ideas already https://github.com/rails/rails/pull/37944#issuecomment-565847607 but as suggested I wouldn't implement it here to keep this PR small and easy to grasp.

Simple scopes like [this](https://github.com/rails/rails/pull/37944/files#diff-f61c8d8a428f9dbae2dc668ca71f894e6f2b5947fb28a6a9339bf2854f89073aR64) could be enough for start maybe.

kbrock commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -303,6 +303,151 @@ def _select!(*fields) # :nodoc:
+    #   # )
+    #   # SELECT * FROM posts JOIN posts_with_tags ON posts_with_tags.id = posts.id
+    #
+    # You can also pass the expression as a `String`.
```
If this only accepted Arel, they could pass in an Arel quoted string.

Hoping most people would be able to get this without documentation.

vlado replied:
> Is this really necessary? If you start dropping to raw sql, is there really lots of value over `find_by_sql`.

Value over `find_by_sql` is that it returns `ActiveRecord::Relation` instead of an Array. My line of thinking was that if you have a complex SQL in a string you should be able to pass it as a CTE expression.

```
COMPLEX_QUERY = <<-SQL
    ... very complex query getting and calculating stuff about posts directly in the database ...
SQL

Post.with(popular_posts: COMPLEX_QUERY).from("popular_posts AS posts")
```

This could lead to security problems so I dropped support for plain strings. It is still possible to pass Arel quoted string though

```
Post.with(popular_posts: Arel.sql(COMPLEX_QUERY)).from("popular_posts AS posts")
```

but didn't document it not to mention Arel API. Not sure how to communicate how we should communicate that properly?

vlado replied:
I just checked [documentation](https://api.rubyonrails.org/classes/ActiveRecord/QueryMethods.html#method-i-order) for `order` method and it mentions `Arel.sql` so I suppose we can document it here also

<img width="818" alt="Screenshot 2022-06-27 at 10 22 31" src="https://user-images.githubusercontent.com/5148/175894036-1ede1a2b-440a-41f3-914c-bcf65d93e673.png">.

simi commented:
Btw. CTE can be used internally to batch upsert data (even for has_many through relations). I can try to dive into that once merged.

https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/activerecord/lib/active_record/association_relation.rb#L22

vlado commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -303,6 +303,151 @@ def _select!(*fields) # :nodoc:
+    # or pass them as a `Hash` but be aware that `ActiveRecord::Relation` does not have `UNION` support
+    # so you will probably need to pass the expression as `String` or raw Arel expression.
+    #
+    #   union = non_recursive_relation.arel.union(recursive_relation.arel)
+    #   Comment.with_recursive(top_comment: top_comment, replies: union)
```
@casperisfine I agree with previous comments not to encourage Arel and to drop raw SQL support. Problem here is that ActiveRecord does not have UNION support so you can only pass Arel expression or string with raw SQL.

Maybe we should drop support for recursive CTE queries until UNION support is added?

casperisfine replied:
In general, the smaller the scope, the easier it is to review and merge PR. So first adding support for CTE without recursion would be a good way to move forward.

simi replied:
I agree on this. We can start with simple CTE. I can take a look at introducing UNION to AR and later this can be added.

vlado replied:
Support for recursive queries removed, plan is to add it when UNION support is there.

vlado commented:
@casperisfine @kbrock @simi I've followed your feedback and made following changes:

* Removed all mentions of Arel API from Changelogs and docs.
* Removed support for recursive queries.
* Only supports hash as an argument now. (It is also possible to pass already prepared `Arel::Nodes::As` or `Arel::Nodes::TableAlias` as arguments but not mentioned in the documentation. Not sure if we should remove them also, they could be useful for internal use?)

This is now a minimal implementation to make CTE in Rails easy to use (pass them as `AR::Relation` and get `AR::Relation`back) and to build on in the future.

vlado commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,38 @@
+    *Vlado Cingel*
+
+*   Support passing record to uniqueness validator `:conditions` callable:
+
```
Removed, Thanks!

vlado commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,38 @@
+    posts_with_comments_table = Arel::Table.new(:posts_with_comments)
+    posts_with_comments_expression = Post.where("comments_count > ?", 0).arel
+    Post.all.arel.with(Arel::Nodes::As.new(posts_with_comments_table, posts_with_comments_expression))
+    # => Arel::SelectManager
```
Arel not mentioned in Changelogs and Docs any more.

byroot reviewed (approved):
Thanks for reducing the scope. It looks good to me now.

byroot commented:
Could you rebase? There's lots of unrelated CI failures because of the minitest update. I'll leave a bit of time for other maintainers to chime in if they want, but I don't see any objections to merging.

matthewd commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -1514,6 +1572,35 @@ def build_select(arel)
+        hash.map do |name, value|
+          expression =
+            case value
+            when Arel::Nodes::SqlLiteral then Arel.sql("(#{value})")
```
Use a Grouping node to add parens, don't re-create the SqlLiteral by modifying its string form.

matthewd commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -1514,6 +1572,35 @@ def build_select(arel)
+            else
+              raise ArgumentError, "Unsupported argument type: `#{value}` #{value.class}"
+            end
+          Arel::Nodes::TableAlias.new(expression, Arel.sql(name.to_s))
```
I don't know the specific node this should be off the top of my head, but it definitely should not inject the hash key as raw SQL; `name` should end up table/column quoted.

vlado replied:
Thanks for catching those! Both fixed in https://github.com/rails/rails/pull/37944/commits/5b8676c3c230a3d3c2deeba07cce0949636e0132

vlado commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -1514,6 +1572,35 @@ def build_select(arel)
+
+        with_statements = with_values.map do |with_value|
+          case with_value
+          when Arel::Nodes::As, Arel::Nodes::TableAlias then with_value
```
Agreed, will remove it.

casperisfine replied:
> Removed in https://github.com/rails/rails/commit/464d5ac6f30b1de34f0b6447e3b007a0cbf740dd

Ah forgot to mention, you'll have to squash all these 24 commits into a single one.

vlado replied:
> Ah forgot to mention, you'll have to squash all these 24 commits into a single one.

Sure. Just addressed feedback from @matthewd and will squash if all is ok.

vlado replied:
@casperisfine Squashed and tests are green.

casperisfine commented:
Also the CI failures look legit now.

byroot commented:
Thank you @vlado. Feel free to open a second PR for recursive support.

vlado commented:
> Thank you @vlado. Feel free to open a second PR for recursive support.

Great, thanks @byroot!

nhorton commented:
Just an unsolicited comment on this thread for posterity on the syntax and need for recursive CTEs.
1. I have had a need for recursion somewhere in most non-trivial companies I have ever done as we have ended up with something that is inherently nested. Ex. permssions groups where a group can be a group member, org charts, directed acyclic graphs representing business rules, etc. It is very needed.
2. `with.recursive` feels very much like what I would expect. I think it is very similar conceptually to `where.not`.
